src/protocol/attribute.rs

```diff
@@ -0,0 +1,224 @@
+use super::{AttributionModel, RecordId};
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::mesh::{Gateway, Mesh},
+    protocol::context::ProtocolContext,
+    secret_sharing::Replicated,
+};
+use futures::future::try_join_all;
+
+/// Obilivious attribution protocol
+///
+/// Obliviously attribute trigger events to corresponding source events for different breakdowns in a MPC.
+/// There are several common attribution models such as Last Touch, Even Credit and Exponential Decay.
+/// https://github.com/patcg-individual-drafts/ipa/blob/main/IPA-End-to-End.md#oblivious-last-touch-attribution
+#[allow(dead_code)]
+pub struct Attribute<'a, G> {
+    ctx: &'a ProtocolContext<'a, G, AttributionModel>,
+    step: AttributionModel,
```

Comment:
This isn't going to work, I don't think.  The step needs to be taken from the same global enum.  So in order to have this work, you need to take the top-level enum we're using.  Which means parameterizing on `S: Step` and taking an argument of that type.

I also think that we will probably want to make the implementation of attribution a generic parameter.  That is, you might define a generic attribution step as a trait, then a concrete implementation of that would be last touch.

Reply:
Yeah, I'd recommend just focusing on this one attribution algorithm in isolation for now. Once we have it running, we can think about how to support other stuff. I'd call this module "LastTouchAttribution"